### PR TITLE
refactor: collaboration api v2

### PR DESCRIPTION
**Pull request by 1yefuwang1**
closes https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_sprints/taskboard/Teams%20Extensibility%20E2E%20Team/Microsoft%20Teams%20Extensibility/CY21-11.2?workitem=10951279

### CODE REVIEW

File: packages/fx-core/src/plugins/resource/utils4v2.ts [hunk 1]
```diff
@@ -435,3 +441,37 @@ export function setLocalSettingsV1(pluginContext: PluginContext, localSettings:
...
+  }
+  const pluginContext: PluginContext = convert2PluginContext(plugin.name, ctx, inputs);
+  pluginContext.azureAccountProvider = tokenProvider.azureAccountProvider;
+  pluginContext.appStudioToken = tokenProvider.appStudioToken;
+  pluginContext.graphTokenProvider = tokenProvider.graphTokenProvider;
+  pluginContext.envInfo = newEnvInfo();
+  pluginContext.envInfo.state = flattenConfigMap(state);
+  pluginContext.envInfo.config = envInfo.config as EnvConfig;
+  pluginContext.config = pluginContext.envInfo.state.get(plugin.name) ?? new ConfigMap();
+  const result = await fn(pluginContext, userInfo);
```
Comment (SLdragon) [inline packages/fx-core/src/plugins/resource/utils4v2.ts:475]: Seems we can `return fn(pluginContext, userInfo)` directly without await

File: packages/fx-core/src/plugins/solution/fx-solution/v2/adaptor.ts [hunk 2]
```diff
@@ -107,3 +107,36 @@ export class ProvisionContextAdapter extends BaseSolutionContextAdaptor {
...
 }
+
+export class CollaboratorContextAdapter extends BaseSolutionContextAdaptor {
+  constructor(params: Parameters<NonNullable<v2.SolutionPlugin["grantPermission"]>>) {
+    super();
+    const v2context: v2.Context = params[0];
+    const inputs: Inputs = params[1];
+    const envInfo = params[2];
+    const tokenProvider = params[3];
+    this.root = inputs.projectPath ?? "";
```
Comment (SLdragon) [inline packages/fx-core/src/plugins/solution/fx-solution/v2/adaptor.ts:118]: What will happen when root is empty?

**Comment by SLdragon** [comment]
Some questions related to v1 and v2:
What will we do for v1 logic in the future, do we have time line to remove v1?
And for future development, do we need to implement both v1 and v2 logic?

File: packages/fx-core/src/plugins/solution/fx-solution/v2/adaptor.ts (hunk 2 above)
Comment (1yefuwang1) [reply packages/fx-core/src/plugins/solution/fx-solution/v2/adaptor.ts:118]: root should never be empty here.  What I want here is just to cast away the null-ness of projectPath for simplicity. I'll change to throw error to enforce the fact the root will not be undefined.

**Comment by 1yefuwang1** [comment]
> Some questions related to v1 and v2:
> 
> * What will we do for v1 logic in the future, do we have time line to remove v1?
> * And for future development, do we need to implement both v1 and v2 logic?
I think I have refactored collaboration related code in a V1-V2 agnostic manner as much as possible. Most of the code only depends on what it needs, not just v1.PluginContext, except for code that calls plugin APIs which is different between V1 and V2.
Now,  both v1 and v2 implementations serve just as shims, which, in the end, call into (mostly) V1-V2 agnostic code.
So, I think maintaining both v1 and v2 will not be a problem.

**Review by SLdragon — APPROVED** [review]

**Review by jayzhang — DISMISSED** [review]

**Review by jayzhang — APPROVED** [review]

**Review by LongOddCode — APPROVED** [review]

**Review by SLdragon — APPROVED** [review]

**Pull request merged**